File: 11 semester/Games/LR_5/Source/LR_5/EmptyCellState.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "EmptyCellState.h"

EmptyCellState::EmptyCellState() : CellState()
{
}

EmptyCellState::~EmptyCellState()
{
}

int EmptyCellState::Click() {
    if (IsMark) {
        return -2;
    }
    Update(this);
    return GetBombCount();
}

int EmptyCellState::GetBombCount() {
    int count = 0;
    for (auto cell : Cells) {
        if (cell->IsBomb()) {
            count++;
        }
    }
    return count;
}

void EmptyCellState::Update(IObservable* subject) {
    if (!IsOpen) {
        IsMark = false;
        Open();
        if (GetBombCount() == 0) {
            Notify();
        }
    }
}

File: 11 semester/Games/LR_5_unity/Assets/Scripts/EmptyCellState.cs
public class EmptyCellState : CellState
{
    public override int Click()
    {
        if (IsMark)
        {
            return -2;
        }
        Update(this);
        return GetBombCount();
    }

    public override int GetBombCount()
    {
        int count = 0;
        foreach (var cell in Cells)
        {
            if (cell.CellState is BombCellState)
            {
                count++;
            }
        }
        return count;
    }

    public override void Update(IObservable subject)
    {
        {
            if (IsOpen == false)
            {
                IsMark = false;
                Open();
                if (GetBombCount() == 0)
                {
                    Notify();
                }
            }
        }
    }
}